proposals/lip-0054.md

```diff
@@ -427,6 +427,18 @@ def execute(trs: Transaction) -> None:
         eventsDefaultTopic = ccmID
 
         ccm = decode(crossChainMessageSchema, ccmBytes)
+        ccm.status = CCM_STATUS_CODE_RECOVERED
+        ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID
+
+        try:
+            # Modules can verify the CCM.
+            # The Token module verifies the escrowed balance in the CCM sending chain for the message fee.
+            for each module mdl for which verifyCrossChainMessage exists:
+                mdl.verifyCrossChainMessage(trs, ccm)
+        except:
+            emitCCMEvent(ccm, CCM_PROCESSED_RESULT_DISCARDED, CCM_PROCESSED_CODE_INVALID_CCM_VERIFY_CCM_EXCEPTION)
+            return
```

Comment:
I would not extract this code from the functions. This return statement is wrong, for instance

Reply:
Ok this return should be removed. The intention is not to introduce protocol change, just to avoid using the same code twice. 

I could withdraw this change, no problem from my side.

Reply:
Maybe my point is that we should keep refactoring changes to a minimum to avoid introducing bugs inadvertently. IMHO the current version is better too, I understand there is code duplication, but on the other hand there is a nice separation in each one of those functions that make the code nicely readable.

Reply:
I see your point. I thought it is just a shortcut refactoring, but indeed it is not as simple as that and it does not make sense to spend effort to make it equivalent.

Concretely, in the common execution of those two functions, we have the following block:

```
    ccm.status = CCM_STATUS_CODE_RECOVERED
    ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID

    try:
        # Modules can verify the CCM.
        # The Token module verifies the escrowed balance in the CCM sending chain for the message fee.
        for each module mdl for which verifyCrossChainMessage exists:
            mdl.verifyCrossChainMessage(trs, ccm)
    except:
        emitCCMEvent(ccm, CCM_PROCESSED_RESULT_DISCARDED, CCM_PROCESSED_CODE_INVALID_CCM_VERIFY_CCM_EXCEPTION)
        return
```

In case of exception, then:

i) the rest of the function is not executed and 
ii) the ccm is appended to recovered CCMS.

By moving the common part in the `execute` function, then it is not straighforward to achieve the same outcome. The current version does (ii) but not (i) (i.e., the functions `applyRecovery` or `forwardRecovery` are erroneously called). On the other hand, if we just add `continue` after emitting the event, then we would achieve (i) but not (ii), i.e., we will not append the ccm to the `recoveredCCMs`.  The suggested way to fix it would be to do the following:

```
 for ccmBytes in trsParams.crossChainMessages:
        ccmID = sha256(ccmBytes)
        # Set ccmID as default topic to all events emitted in applyRecovery and forwardRecovery
        # (instead of the transaction ID). The default topic is added automatically by emitEvent.
        eventsDefaultTopic = ccmID

        ccm = decode(crossChainMessageSchema, ccmBytes)
        ccm.status = CCM_STATUS_CODE_RECOVERED
        ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID

        try:
            # Modules can verify the CCM.
            # The Token module verifies the escrowed balance in the CCM sending chain for the message fee.
            for each module mdl for which verifyCrossChainMessage exists:
                mdl.verifyCrossChainMessage(trs, ccm)

            # If the sending chain is the mainchain, recover the CCM.
            # This function never raises an error.
            if ccm.sendingChainID == getMainchainID():
                applyRecovery(trs, ccm)
            # If the sending chain is not the mainchain, forward the CCM.
            # This function never raises an error.
            elif ccm.sendingChainID != getMainchainID():
                forwardRecovery(trs, ccm)

        except:
            emitCCMEvent(ccm, CCM_PROCESSED_RESULT_DISCARDED, CCM_PROCESSED_CODE_INVALID_CCM_VERIFY_CCM_EXCEPTION)        
       
        # Append the recovered CCM to the list of recovered CCMs.
        # Notice that the ccm has been mutated in the applyRecovery and forwardRecovery functions
        # as the status is set to CCM_STATUS_CODE_RECOVERED (so that it cannot be recovered again).
        recoveredCCMs.append(encode(crossChainMessageSchema, ccm))    

```

This I think would work and have the intended behavior.

However it makes indeed the `execute` function more complicated (so overall we don't even make specs simpler) and introduces the un-necessary extra effort to check correctness compared to the previous version.  I would therefore recommend to revert this change and keep the current specs.

Reply:
I reverted this change, since this seemed to be the most reasonable option.